share_app/src/main/java/org/odk/share/tasks/DownloadJob.java

```diff
@@ -352,6 +355,39 @@ private String receiveFile(String path) {
             filename = dis.readUTF();
             long fileSize = dis.readLong();
             Timber.d("Size of file " + filename + " " + fileSize);
+
+            int x = 0;
+            boolean saved = false;
+            while (!saved && x < 3) {
+                saved = persistFile(path, filename, fileSize, x);
+                x++;
+            }
+        } catch (Exception e) {
+            Timber.e(e);
+        }
+        return filename;
+    }
+
+    /**
+     * Fix issue on some android phones loosing permission to write to disk
+     * Happened on android lollipop
+     **/
+    private boolean persistFile(String path, String filename, long fileSize, int retry) throws Exception {
+
+        if (retry > 0) {
+            if (ContextCompat.checkSelfPermission(getContext().getApplicationContext(), Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                    != PackageManager.PERMISSION_GRANTED) {
+                // Permission lost, retry after 2 seconds then retry
+                System.gc();
+                try {
+                    Thread.sleep(1000);
```

Comment:
@rkodev Is access being revoked temporarily something other devs have experienced? I guess sleeping for a second will work but we probably should look for a more elegant solution

Reply:
I haven't seen it before and It was extremely difficult to recreate. As mentioned before it happened only twice in very many attempts to recreate the issue > 25 times. 

This was the closest issue that I found from another developer.
https://stackoverflow.com/questions/23527767/open-failed-eacces-permission-denied

The very brief explanation was that the method FileOutputStream was being called to many times too frequently.  That's what informed my decision to sleep the thread.

Reply:
@rkodev Cool. I see a FileOutputStream object is created for every file (which I think is fine). Could it be how we're handing the buffer? A few pointers:
 - What's the effect of changing the buffer size? Is there an optimal size?
 - Are we supposed to flush the buffer before closing the output stream?